libraries/chain/controller.cpp

```diff
@@ -628,12 +633,52 @@ struct building_block {
                         v);
    }
 
+   digest_type get_finality_mroot_claim(const block_state& parent, const qc_claim_t& qc_claim) {
+      auto next_core_metadata = parent.core.next_metadata(qc_claim);
+
+      // For proper IF blocks that do not have an associated Finality Tree defined
+      if (parent.core.is_genesis_block_num(next_core_metadata.final_on_strong_qc_block_num)) {
+         return digest_type{};
+      }
+
+      return parent.get_validation_mroot(next_core_metadata.final_on_strong_qc_block_num);
+   }
+
+   qc_data_t get_qc_data(fork_database& fork_db, const block_state& parent) {
+      // find most recent ancestor block that has a QC by traversing fork db
+      // branch from parent
+
+      return fork_db.apply_s<qc_data_t>([&](const auto& forkdb) {
+         auto branch = forkdb.fetch_branch(parent.id());
+
+         for( auto it = branch.begin(); it != branch.end(); ++it ) {
+            if( auto qc = (*it)->get_best_qc(); qc ) {
+               EOS_ASSERT( qc->block_num <= block_header::num_from_id(parent.id()), block_validate_exception,
+                           "most recent ancestor QC block number (${a}) cannot be greater than parent's block number (${p})",
+                           ("a", qc->block_num)("p", block_header::num_from_id(parent.id())) );
+               auto qc_claim = qc->to_qc_claim();
+               if( parent.is_needed(qc_claim) ) {
+                  return qc_data_t{ *qc, qc_claim };
+               } else {
+                  // no new qc info, repeat existing
+                  return qc_data_t{ {},  parent.core.latest_qc_claim() };
+               }
+            }
+         }
+
+         // This only happens when parent block is the IF genesis block or starting from snapshot.
+         // There is no ancestor block which has a QC. Construct a default QC claim.
+         return qc_data_t{ {}, parent.core.latest_qc_claim() };
+      });
+   }
```

Comment:
I think these two should be `const` member functions of `block_state`.

Reply:
I don't think `block_state` should be aware of `fork_database`.

Reply:
I agree. We should not make `block_state` aware of fork_db for `get_qc_data`. We could move `get_finality_mroot_claim` to `block_state`. But it is only used here. It might be clearer to implement  it along with `get_qc_data`.

Reply:
> I don't think `block_state` should be aware of `fork_database`.

otoh, I don't think `controller.cpp` should be 5,000 lines long :-).
Because `fork_database` is the way to resolve the parent id of a `block_state`, not allowing `block_state` to use `fork_database` seriously limits what it can do.

Reply:
In 6.1, we should consider splitting controller.cpp.

Reply:
I feel that just splitting it up into two files, but keeping the same organization would not help. I really feel that we should avoid free functions like the ones I pointed to above, and have these be members of the class that makes most sense. I feel this is a more `C++` way of coding, versus having data (struct) definitions and free functions like is done in `C`.

Reply:
I managed to move the two functions to block_state without exposing forks_db to it.